From acec57c77e10230fc32820046301cb2a59c431d4 Mon Sep 17 00:00:00 2001
From: Stefan Agner <stefan@agner.ch>
Date: Thu, 17 Aug 2023 17:38:13 +0200
Subject: [PATCH] board: green: add Green board to read board specific EEPROM

Use board specific code to read serial number and MAC address from
on-board I2C EEPROM.

Signed-off-by: Stefan Agner <stefan@agner.ch>
---
 arch/arm/dts/rk3566-ha-green.dts      |  7 +++
 arch/arm/mach-rockchip/rk3568/Kconfig |  7 +++
 board/nabucasa/ha-green/Kconfig       | 15 +++++++
 board/nabucasa/ha-green/Makefile      |  3 ++
 board/nabucasa/ha-green/ha-green.c    | 65 +++++++++++++++++++++++++++
 configs/green_defconfig               |  3 ++
 include/configs/ha-green.h            | 11 +++++
 7 files changed, 111 insertions(+)
 create mode 100644 board/nabucasa/ha-green/Kconfig
 create mode 100644 board/nabucasa/ha-green/Makefile
 create mode 100644 board/nabucasa/ha-green/ha-green.c
 create mode 100644 include/configs/ha-green.h

diff --git a/arch/arm/dts/rk3566-ha-green.dts b/arch/arm/dts/rk3566-ha-green.dts
index ea9d358099..91afadf0e6 100644
--- a/arch/arm/dts/rk3566-ha-green.dts
+++ b/arch/arm/dts/rk3566-ha-green.dts
@@ -449,6 +449,13 @@
 		reg = <0x51>;
 		#clock-cells = <0>;
 	};
+
+	eeprom: eeprom@50 {
+		compatible = "atmel,24c32";
+		reg = <0x50>;
+		pagesize = <32>;
+	};
+
 };
 
 &i2s0_8ch {
diff --git a/arch/arm/mach-rockchip/rk3568/Kconfig b/arch/arm/mach-rockchip/rk3568/Kconfig
index baa51349f4..42652dec03 100644
--- a/arch/arm/mach-rockchip/rk3568/Kconfig
+++ b/arch/arm/mach-rockchip/rk3568/Kconfig
@@ -17,6 +17,12 @@ config TARGET_ANBERNIC_RGXX3_RK3566
 	  and RG503. The correct device tree name will automatically
 	  be selected by the bootloader.
 
+config TARGET_NABU_CASA_HA_GREEN_RK3566
+	bool "Nabu Casa Home Assistant Green"
+	help
+	  Nabu Casa's Home Assistant Green single board computer with a RK3566
+	  SoC.
+
 config TARGET_ODROID_M1_RK3568
 	bool "ODROID-M1"
 	help
@@ -44,6 +50,7 @@ config SYS_MALLOC_F_LEN
 source "board/rockchip/evb_rk3568/Kconfig"
 source "board/anbernic/rgxx3_rk3566/Kconfig"
 source "board/hardkernel/odroid_m1/Kconfig"
+source "board/nabucasa/ha-green/Kconfig"
 source "board/pine64/quartz64_rk3566/Kconfig"
 
 endif
diff --git a/board/nabucasa/ha-green/Kconfig b/board/nabucasa/ha-green/Kconfig
new file mode 100644
index 0000000000..863afd4363
--- /dev/null
+++ b/board/nabucasa/ha-green/Kconfig
@@ -0,0 +1,15 @@
+if TARGET_NABU_CASA_HA_GREEN_RK3566
+
+config SYS_BOARD
+    default "ha-green"
+
+config SYS_VENDOR
+    default "nabucasa"
+
+config SYS_CONFIG_NAME
+    default "ha-green"
+
+config BOARD_SPECIFIC_OPTIONS # dummy
+    def_bool y
+
+endif
diff --git a/board/nabucasa/ha-green/Makefile b/board/nabucasa/ha-green/Makefile
new file mode 100644
index 0000000000..bf711fe8c4
--- /dev/null
+++ b/board/nabucasa/ha-green/Makefile
@@ -0,0 +1,3 @@
+# SPDX-License-Identifier: GPL-2.0+
+
+obj-y    += ha-green.o
diff --git a/board/nabucasa/ha-green/ha-green.c b/board/nabucasa/ha-green/ha-green.c
new file mode 100644
index 0000000000..b9a0c32441
--- /dev/null
+++ b/board/nabucasa/ha-green/ha-green.c
@@ -0,0 +1,65 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright 2023 Nabu Casa, Inc.
+ * Copyright 2022 Google LLC
+ */
+#include <net.h>
+#include <errno.h>
+#include <dm/device.h>
+#include <dm/uclass.h>
+#include <common.h>
+#include <dm.h>
+#include <eeprom.h>
+#include <env.h>
+#include <i2c_eeprom.h>
+#include <netdev.h>
+
+#define CONTENT_SIZE 44
+#define ETHADDR_OFFSET 0X22
+#define ETHADDR_LEN 6
+#define SERIALNUM_OFFSET 0X10
+#define SERIALNUM_LEN 18
+#define VERSION_OFFSET 0X0C
+#define VERSION_LEN 4
+
+int misc_init_r(void)
+{
+	unsigned char data[CONTENT_SIZE];
+	unsigned char ethaddr[ETHADDR_LEN];
+	unsigned char serialnum[SERIALNUM_LEN + 1];
+	unsigned char version[VERSION_LEN];
+
+	struct udevice *dev;
+	int ret;
+
+	ret = uclass_first_device_err(UCLASS_I2C_EEPROM, &dev);
+	if (ret) {
+		printf("UCLASS_I2C_EEPROM get error\n");
+		return 0;
+	}
+
+	ret = i2c_eeprom_read(dev, 0, data, sizeof(data));
+	if (ret) {
+		printf("UCLASS_I2C_EEPROM read error %d\n", ret);
+		return 0;
+	}
+
+	if (!(data[0] == 3 && data[1] == 5 && data[2] == 6 && data[3] == 6))
+	{
+		printf("EEPROM header not ok\n");
+		return 0;
+	}
+
+	memcpy(version , data + VERSION_OFFSET, VERSION_LEN);
+	printf("Board revision: %c%c\n", version[2], version[3]);
+
+	memcpy(serialnum, data + SERIALNUM_OFFSET, SERIALNUM_LEN);
+	serialnum[SERIALNUM_LEN] = 0;
+	env_set("serial#", serialnum);
+
+	memcpy(ethaddr, data + ETHADDR_OFFSET, ETHADDR_LEN);
+	if (is_valid_ethaddr(ethaddr))
+		eth_env_set_enetaddr("ethaddr", ethaddr);
+
+	return 0;
+}
diff --git a/configs/green_defconfig b/configs/green_defconfig
index 5f94c06279..a97391f295 100644
--- a/configs/green_defconfig
+++ b/configs/green_defconfig
@@ -16,6 +16,7 @@ CONFIG_SPL_ROCKCHIP_COMMON_BOARD=y
 CONFIG_ROCKCHIP_SPI_IMAGE=y
 CONFIG_SPL_SERIAL=y
 CONFIG_SPL_STACK_R_ADDR=0x600000
+CONFIG_TARGET_NABU_CASA_HA_GREEN_RK3566=y
 CONFIG_SPL_STACK=0x400000
 CONFIG_DEBUG_UART_BASE=0xFE660000
 CONFIG_DEBUG_UART_CLOCK=24000000
@@ -60,6 +61,8 @@ CONFIG_SPL_CLK=y
 CONFIG_ROCKCHIP_GPIO=y
 CONFIG_SYS_I2C_ROCKCHIP=y
 CONFIG_MISC=y
+CONFIG_I2C_EEPROM=y
+CONFIG_SYS_I2C_EEPROM_ADDR=0x0
 CONFIG_SUPPORT_EMMC_RPMB=y
 CONFIG_MMC_DW=y
 CONFIG_MMC_DW_ROCKCHIP=y
diff --git a/include/configs/ha-green.h b/include/configs/ha-green.h
new file mode 100644
index 0000000000..e42d9c8e8c
--- /dev/null
+++ b/include/configs/ha-green.h
@@ -0,0 +1,11 @@
+/* SPDX-License-Identifier: GPL-2.0+ */
+
+#ifndef __HA_GREEN_H
+#define __HA_GREEN_H
+
+#define ROCKCHIP_DEVICE_SETTINGS \
+    "cramfsaddr=0x0c000000\0"
+
+#include <configs/rk3568_common.h>
+
+#endif
-- 
2.43.0

